util.cpp

```diff
@@ -0,0 +1,55 @@
+#include <stdexcept>
+#include <boost/multiprecision/cpp_int.hpp>
+#include "include/util.h"
+
+using uint128_t = boost::multiprecision::uint128_t;
+
+const ull ULLMAX = std::numeric_limits<ull>::max();
+
+unsigned long long int double_to_ull(double d, double epsilon) {
+  return (unsigned long long) (d + epsilon);
+}
+
+ull nondirectional_non_self_edge_pairing_fn(ull i, ull j) {
+  // swap i,j if necessary
+  if (i > j) {
+    i^=j;
+    j^=i;
+    i^=j;
+  }
+  ull jm = j-1ull;
```

Comment:
maybe there's something subtle here that I'm not seeing but isn't all of this equivalent to

const ull max_in = std::max(i, j);
const ull min_in = std::min(i, j);
return min_in * max_in * (max_in / 2ull)

If it is, I think the above will be much faster given that there's no branching.

Reply:
We're potentially working with numbers where these calculations could silently overflow. We want to test if any of these operations would overflow, and report it.

Reply:
I think that can still be tested though (however this is fine)